### ISSUE TITLE
Debug issues with Topology/Systems having different number of atoms

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#140]
Just ran into this issue in `testsystem` with the myb system:
```
================================================================================
SAMS sampler iteration     1
--------------------------------------------------------------------------------
Expanded Ensemble sampler iteration        1
................................................................................
MCMC sampler iteration 1
Taking 500 steps of GHMC...
Final energy is    -5428.303 kT
................................................................................
Updating chemical state with ncmc-geometry-ncmc scheme...
Proposing new topology...
Proposed transformation: GLY-VAL-ARG-LYS-GLY-TRP-HIS-GLU-HIS-VAL-THR-GLN-ASP-LEU-ARG-SER-HIS-LEU-VAL-HIS-LYS-LEU-VAL-GLN-ALA-ILE-PHE-PRO-THR-PRO-ASP-PRO-ALA-ALA-LEU-LYS-ASP-ARG-ARG-MET-GLU-ASN-LEU-VAL-ALA-TYR-ALA-LYS-LYS-VAL-GLU-GLY-ASP-MET-TYR-GLU-SER-ALA-ASN-SER-ARG-ASP-GLU-TYR-TYR-HIS-LEU-LEU-ALA-GLU-LYS-ILE-TYR-LYS-ILE-GLN-LYS-GLU-LEU-GLU-GLU-LYS-ARG-ARG-SER-ARG-LEU-LYS-GLU-LYS-ARG-ILE-LYS-GLU-LEU-GLU-LEU-LEU-LEU-MET-SER-THR-GLU-ASN-GLU-LEU-LYS-GLY-GLN-GLN-ALA-LEU => GLY-VAL-ARG-LYS-GLY-TRP-HIS-GLU-HIS-VAL-THR-GLN-ASP-LEU-ARG-SER-HIS-THR-VAL-HIS-LYS-LEU-VAL-GLN-ALA-ILE-PHE-PRO-THR-PRO-ASP-PRO-ALA-ALA-LEU-LYS-ASP-ARG-ARG-MET-GLU-ASN-LEU-VAL-ALA-TYR-ALA-LYS-LYS-VAL-GLU-GLY-ASP-MET-TYR-GLU-SER-ALA-ASN-SER-ARG-ASP-GLU-TYR-TYR-HIS-LEU-LEU-ALA-GLU-LYS-ILE-TYR-LYS-ILE-GLN-LYS-GLU-LEU-GLU-GLU-LYS-ARG-ARG-SER-ARG-LEU-LYS-GLU-LYS-ARG-ILE-LYS-GLU-LEU-GLU-LEU-LEU-LEU-MET-SER-THR-GLU-ASN-GLU-LEU-LYS-GLY-GLN-GLN-ALA-LEU
Generating new system...
Performing NCMC annihilation
LOG: Creating alchemically modified system...
LOG: particle 177 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 263 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 480 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 756 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 898 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 924 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 959 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1031 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1052 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1194 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1419 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1723 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1734 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: Elapsed time 0.686 s.
LOG: Creating alchemically modified intermediate...
LOG: Elapsed time 0.349 s.
LOG: NCMC logP      -63.4 | initial_total_energy    -6154.6 kT | final_total_energy    -5996.3 kT.
Geometry engine proposal...
Traceback (most recent call last):
  File "perses/tests/testsystems.py", line 772, in <module>
    testsystem.sams_samplers[solvent + '-complex'].run(niterations=100)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 921, in run
    self.update()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 905, in update
    self.update_sampler()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 873, in update_sampler
    self.sampler.update()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 751, in update
    self.update_state()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 693, in update_state
    geometry_logp_reverse = self.geometry_engine.logp_reverse(topology_proposal, geometry_new_positions, geometry_old_positions, self.sampler.thermodynamic_state.beta)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 119, in logp_reverse
    logp_proposal, _ = self._logp_propose(top_proposal, old_coordinates, beta, new_positions=new_coordinates, direction='reverse')
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 165, in _logp_propose
    atom_proposal_order, logp_choice = proposal_order_tool.determine_proposal_order(direction='reverse')
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 863, in determine_proposal_order
    structure = parmed.openmm.load_topology(self._topology_proposal.old_topology, self._topology_proposal.old_system)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/parmed/utils/decorators.py", line 31, in new_fcn
    return fcn(*args, **kwargs)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/parmed/openmm/topsystem.py", line 134, in load_topology
    '(%d vs. %d)' % (len(struct.atoms), system.getNumParticles()))
TypeError: Topology and System have different numbers of atoms (1889 vs. 1894)
```
Will debug in the morning.

**Comment by jchodera** [comment]
Actually, it looks like this issue has to do with the `reverse` mode of `GeometryEngine`'s `self._logp_propose`:
```
======================================================================
ERROR: Testing MultiTargetDesign sampler with AlanineDipeptideTestSystem transfer free energy from vacuum -> vacuum
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 1044, in run
    self.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 1026, in update
    self.update_samplers()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 990, in update_samplers
    sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 906, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 874, in update_sampler
    self.sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 752, in update
    self.update_state()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 694, in update_state
    geometry_logp_reverse = self.geometry_engine.logp_reverse(topology_proposal, geometry_new_positions, geometry_old_positions, self.sampler.thermodynamic_state.beta)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 119, in logp_reverse
    logp_proposal, _ = self._logp_propose(top_proposal, old_coordinates, beta, new_positions=new_coordinates, direction='reverse')
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 165, in _logp_propose
    atom_proposal_order, logp_choice = proposal_order_tool.determine_proposal_order(direction='reverse')
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 873, in determine_proposal_order
    structure = parmed.openmm.load_topology(self._topology_proposal.old_topology, self._topology_proposal.old_system)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/parmed/utils/decorators.py", line 31, in new_fcn
    return fcn(*args, **kwargs)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/parmed/openmm/topsystem.py", line 134, in load_topology
    '(%d vs. %d)' % (len(struct.atoms), system.getNumParticles()))
TypeError: Topology and System have different numbers of atoms (31 vs. 22)
-------------------- >> begin captured logging << --------------------
alchemy.alchemy: DEBUG: Creating alchemically modified system...
alchemy.alchemy: DEBUG: Elapsed time 0.004 s.
alchemy.alchemy: DEBUG: Creating alchemically modified intermediate...
alchemy.alchemy: DEBUG: Elapsed time 0.002 s.
root: DEBUG: NCMC logP       -3.5 | initial_total_energy      +73.2 kT | final_total_energy      +82.1 kT.
--------------------- >> end captured logging << ---------------------
```

**Comment by jchodera** [comment]
I'm still seeing this error:
https://travis-ci.org/choderalab/perses/jobs/111885576#L5883
```
======================================================================
FAIL: Doctest: perses.samplers.samplers.SAMSSampler
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/_test/lib/python2.7/doctest.py", line 2226, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for perses.samplers.samplers.SAMSSampler
  File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 786, in SAMSSampler
----------------------------------------------------------------------
File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 830, in perses.samplers.samplers.SAMSSampler
Failed example:
    sams_sampler.run()
Exception raised:
    Traceback (most recent call last):
      File "/home/travis/miniconda/envs/_test/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest perses.samplers.samplers.SAMSSampler[12]>", line 1, in <module>
        sams_sampler.run()
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 922, in run
        self.update()
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 906, in update
        self.update_sampler()
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 874, in update_sampler
        self.sampler.update()
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 752, in update
        self.update_state()
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 694, in update_state
        geometry_logp_reverse = self.geometry_engine.logp_reverse(topology_proposal, geometry_new_positions, geometry_old_positions, self.sampler.thermodynamic_state.beta)
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 122, in logp_reverse
        logp_proposal, _ = self._logp_propose(top_proposal, old_coordinates, beta, new_positions=new_coordinates, direction='reverse')
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 168, in _logp_propose
        atom_proposal_order, logp_choice = proposal_order_tool.determine_proposal_order(direction='reverse')
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 863, in determine_proposal_order
        structure = parmed.openmm.load_topology(self._topology_proposal.old_topology, self._topology_proposal.old_system)
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/parmed/utils/decorators.py", line 31, in new_fcn
        return fcn(*args, **kwargs)
      File "/home/travis/miniconda/envs/_test/lib/python2.7/site-packages/parmed/openmm/topsystem.py", line 134, in load_topology
        '(%d vs. %d)' % (len(struct.atoms), system.getNumParticles()))
    TypeError: Topology and System have different numbers of atoms (31 vs. 22)
```
Any idea what might be going on?

**Comment by jchodera** [comment]
Fixed thanks to #158!
